File: server/modules/tool-handler.mjs
/**
 * AI Travel Butler - Tool Call Handler
 * 处理来自AI模型的工具调用
 */

import { normalizeLocation, normalizeRecommendations, parseToolArguments } from './normalizer.mjs';

/**
 * 解析AI返回的工具调用结果
 */
export function mapToolCalls(toolCalls, mcpTrace, provider) {
  const dayPlanItinerary = [];
  let socialRecommendations = [];

  console.log(`[DEBUG] Processing ${toolCalls.length || 0} tool calls from ${provider}:`);
  
  for (const tc of toolCalls || []) {
    console.log(`  [RAW CALL] Name: ${tc.name || tc.function?.name || 'unknown'}, Args: ${(tc.function && tc.function.arguments) || JSON.stringify(tc.args)}`);
    
    const fnName = tc.function?.name || tc.name;
    const rawArgs = tc.function?.arguments;
    let args;
    
    // 针对DeepSeek的特殊处理 - 如果是get_social_recommendations，直接解析
    if (fnName === 'get_social_recommendations') {
      if (rawArgs) {
        try {
          args = JSON.parse(rawArgs);
        } catch (e) {
          // 针对DeepSeek可能出现的转义问题特别处理
          try {
            // 尝试处理字符串化的参数
            const cleanedArgs = rawArgs
              .replace(/\\\\\\/g, '') // 移除重复的转义斜杠
              .replace(/\\"/g, '"') // 将 \" 转换为 "
              .replace(/\\n/g, '') // 移除换行符
              .replace(/\\r/g, '') // 移除回车符
              .trim();
            
            args = JSON.parse(cleanedArgs);
            
            // 颼果解析成功但recommendations还是字符串，再次处理
            if (args.recommendations && typeof args.recommendations === 'string') {
              try {
                args.recommendations = JSON.parse(args.recommendations);
              } catch (secondaryParse) {
                console.log(`  [INFO] Recommendation data still unparseable, trying basic parse`, args.recommendations);
                // 使用基本正则解析
                const recMatches = args.recommendations.match(/\\{[^}]+\\}/g) || [args.recommendations];
                if (recMatches.length > 0) {
                  const parsedRecs = [];
                  for (const recItem of recMatches) {
                    const titleMatch = recItem.match(/"title"[^"]*"([^"]+)"/);
                    const rankMatch = recItem.match(/"rank"[^0-9]*([0-9]+)/);
                    if (titleMatch && rankMatch) {
                      parsedRecs.push({
                        title: titleMatch[1],
                        rank: parseInt(rankMatch[1])
                      });
                    }
                  }
                  if (parsedRecs.length > 0) {
                    args.recommendations = parsedRecs;
                  }
                }
              }
            }
            console.log(`[DEEPSEEK-FIX] Success parsing special-format social recommendations`);
          } catch (cleaningError) {
            console.error(`[ERROR] DeepSeek argument cleaning failed:`, cleaningError.message);
            args = { recommendations: [] }; // 默认值
          }
        }
      } else {
        args = tc.args || {};
      }
    } else {
      // 处他函数使用通用解析方法
      args = rawArgs ? parseToolArguments(rawArgs) : tc.args || {};
    }

    if (fnName === 'location') {
      console.log(`  [DEBUG] Processing location: ${args.name}, coords: (${args.lat}, ${args.lng}), city: ${args.city}`);
      const item = normalizeLocation(args, provider);
      if (item) {
        dayPlanItinerary.push(item);
        console.log(`  [SUCCESS] Normalized location: ${item.name} in ${item.city}`);
      } else {
        console.log(`  [SKIPPED] Invalid location data: ${JSON.stringify(args)}`);
      }
      mcpTrace.push(`tool:${provider}:location`);
    }

    if (fnName === 'get_social_recommendations') {
      socialRecommendations = normalizeRecommendations(args.recommendations || [], provider);
      console.log(`  [INFO] Normalized ${socialRecommendations.length} social recommendations`);
      console.log(`    Recommendations:`, socialRecommendations.slice(0, 3)); // 打印前3个以避免过多输出
      mcpTrace.push(`tool:${provider}:get_social_recommendations`);
    }
  }
  
  // 添加MCP工具处理逻辑
  const mcpResults = [];

  console.log(`[RESULT] Completed processing. Locations: ${dayPlanItinerary.length}, Social Recs: ${socialRecommendations.length}`);
  return { 
    dayPlanItinerary, 
    socialRecommendations,
    mcpResults 
  };
}

File: python_server/modules/tool_handler.py
"""
AI Travel Butler - Tool Call Handler
"""

import json
import re
from typing import List, Dict, Any, Tuple


def parse_tool_arguments(raw_args: str) -> Dict[str, Any]:
    """Parse tool arguments from JSON string."""
    if not raw_args:
        return {}

    try:
        return json.loads(raw_args)
    except json.JSONDecodeError:
        # Try to fix common JSON issues
        try:
            cleaned = raw_args.replace("\\'", '"').replace("'", '"')
            # Fix: remove newlines after colon before array/object
            cleaned = re.sub(r':\s*\n\s*(\[|\{)', r': \1', cleaned)
            cleaned = re.sub(r"([{,]\s*)([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:", r'\1"\2":', cleaned)
            return json.loads(cleaned)
        except Exception:
            return {}


def normalize_location(args: Dict[str, Any], provider: str) -> Dict[str, Any]:
    """Normalize location data from tool call."""
    name = args.get("name")
    lat = args.get("lat")
    lng = args.get("lng")
    city = args.get("city")

    if not name or not lat or not lng or not city:
        return None

    # Convert string coordinates to float
    try:
        lat_float = float(lat) if isinstance(lat, (int, float, str)) else 39.9042
        lng_float = float(lng) if isinstance(lng, (int, float, str)) else 116.4074
    except (ValueError, TypeError):
        lat_float = 39.9042
        lng_float = 116.4074

    return {
        "name": str(name),
        "lat": lat_float,
        "lng": lng_float,
        "city": str(city),
        "day": args.get("day", 1),
        "sequence": args.get("sequence", 1),
        "time": args.get("time", ""),
        "transit_hint": args.get("transit_hint", ""),
        "description": args.get("description", ""),
        "category": args.get("category", "SIGHT"),
        "tags": args.get("tags", []),
        "address": args.get("address"),
        "rating": args.get("rating"),
        "source_provider": provider
    }


def normalize_recommendations(recommendations: List[Dict], provider: str) -> List[Dict[str, Any]]:
    """Normalize social recommendations."""
    normalized = []
    for rec in recommendations:
        if isinstance(rec, dict):
            normalized.append({
                "name": rec.get("title") or rec.get("name", ""),
                "platform": rec.get("platform", "all"),
                "category": rec.get("category", ""),
                "description": rec.get("reason") or rec.get("description", ""),
                "tags": rec.get("tags", []),
                "rank": rec.get("rank", 0),
                "hot_score": rec.get("hot_score", ""),
                "photo_tips": rec.get("photo_tips", ""),
                "source_provider": provider
            })
    return normalized


def map_tool_calls(tool_calls: List[Dict], mcp_trace: List[str], provider: str) -> Dict[str, Any]:
    """Process tool calls from AI model."""
    day_plan_itinerary = []
    social_recommendations = []

    print(f"[DEBUG] Processing {len(tool_calls) if tool_calls else 0} tool calls from {provider}")

    for tc in tool_calls or []:
        fn_name = tc.get("function", {}).get("name") or tc.get("name", "unknown")
        raw_args = tc.get("function", {}).get("arguments")
        args = {}

        # Special handling for get_social_recommendations
        if fn_name == "get_social_recommendations":
            if raw_args:
                args = parse_tool_arguments(raw_args)
                if not args:
                    args = {"recommendations": []}
                # Handle case where recommendations is still a string
                if isinstance(args.get("recommendations"), str):
                    try:
                        args["recommendations"] = json.loads(args["recommendations"])
                    except Exception:
                        args["recommendations"] = []
            else:
                args = tc.get("args", {})
        else:
            if isinstance(raw_args, dict):
                args = raw_args
            elif isinstance(raw_args, str):
                args = parse_tool_arguments(raw_args)
            else:
                args = tc.get("args", {}) if isinstance(tc.get("args"), dict) else {}

        if fn_name == "location":
            print(f"  [DEBUG] Processing location: {args.get('name')}")
            item = normalize_location(args, provider)
            if item:
                day_plan_itinerary.append(item)
                print(f"  [SUCCESS] Normalized location: {item['name']}")
            mcp_trace.append(f"tool:{provider}:location")

        if fn_name == "get_social_recommendations":
            social_recommendations = normalize_recommendations(args.get("recommendations", []), provider)
            print(f"  [INFO] Normalized {len(social_recommendations)} social recommendations")
            mcp_trace.append(f"tool:{provider}:get_social_recommendations")

    print(f"[RESULT] Locations: {len(day_plan_itinerary)}, Social Recs: {len(social_recommendations)}")
    return {
        "dayPlanItinerary": day_plan_itinerary,
        "socialRecommendations": social_recommendations
    }